### PR TITLE
#6566 refactoring and improvements

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/6566

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+@Repository
+public class ContractPartnerRepository
+{
+	public ImmutableSet<BPartnerId> retrieveAllPartnersWithContracts()
+	{
+		IQueryBL queryBL = Services.get(IQueryBL.class);
+
+		final IQuery<I_C_Flatrate_Term> contractsFilter = queryBL.createQueryBuilder(I_C_Flatrate_Term.class)
+				.addOnlyActiveRecordsFilter()
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java:22]: now i'm wondering, what about e.g. in-progress terms. i think you need to filter only for completed and closed ones

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import de.metas.bpartner.BPartnerId;
+import de.metas.contracts.model.I_C_Flatrate_Term;
+import de.metas.util.Services;
+
+@Repository
+public class ContractPartnerRepository
+{
+	public ImmutableSet<BPartnerId> retrieveAllPartnersWithContracts()
```
Comment (metas-ts) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java:17]:
```suggestion
	public ImmutableSet<BPartnerId> retrieveIdsOfPartnerWithContracts()
```

* include the "Ids" in the name
* in other mf places, "All" implies also inactive, which is not the case here

Comment (metas-ts) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java:17]: There might be 500k bpartners; please return an iterator
(not sure if i approved the set in a previous PR, but i think it's wrong and we have `IQuery.iterateIds(...)`)

**Comment by metas-rc** [comment]
> Pls check my comments.
> We were talking about fixing incorrect masterenddates: shouldn't there be a script for that?

~~As per my comment that I forgot to write~~ Those cases are not affecting the reports after my other changes. I talked with Mark about them and he suggested to only fix them in a followup if they become problematic in the future and if this is visible to the customer so we can budget the fix in another task.
After my last remote test, there are only 13 lines in the report without customer retention for customers that are really no (longer)  customers.

image

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java:22]: Right. Done.

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java:17]: Agree. Making it an iterator.

**Pull request merged**